### PR TITLE
Filter doorhanger using locale grantlist, blocklist, and launch date

### LINKED ISSUE #1961: The DE localization restricted experiments are still available in Test Pilot doorhanger
[Notes]:
- The doorhanger will be removed eventually by #1922, but until then maybe we should not display the experiments restricted for some locales.

[Affected versions]:
- Firefox 50.1.0

[Affected Platforms]:
- All Windows
- All Mac
- All Linux

[Prerequisites]:
- Install or extract a German Firefox build for this issue.
- The "testpilot.env" string pref was manually added with "dev" value on a new clean profile in "about:config" page.
- Test Pilot add-on was previously installed on the profile.

[Steps to reproduce]:
1. Open the DE Firefox build with the profile from prerequisites.
2. Navigate to "https://testpilot.dev.mozaws.net" page and observe the available experiments.
3. Click on the Test Pilot icon from browser tooblar.
4. Observe the available experiments.

[Expected result]:
The DE localization restricted experiments are not displayed on the Test Pilot page or in doorhanger.

[Actual result]:
Step 2. The restricted experiments are not displayed on the page.
Step 4. Activity Stream, Tracking Protection and NoMore 404s experiments are still displayed in doorhanger.

[Additional Notes]:
- Vice versa, if other build is used than DE, Cliqz is available in Test Pilot doorhanger.
- The "New" badge is also displayed for Cliqz experiment on other locales than DE.
- Here is a screenrecording using Firefox DE build: https://goo.gl/TuwaO4.
- Here is a screenrecording using Firefox EN build: https://goo.gl/1jg4kX.

### PR DESCRIPTION
**Pull request by jaredhirsch**
Fixes #1961, #1966, #1967.

### CODE REVIEW

**Comment by jaredhirsch** [comment]
Please don't merge, I'm going to fix 1967 here as well.

File: addon/lib/toolbar-button.js [hunk 1]
```diff
@@ -68,9 +69,22 @@ function getExperimentList(availableExperiments, installedAddons) {
...
+  .filter(x => {
+    if (!('locale_grantlist' in x)) {
+      return true;
+    }
+
+    // In keeping with the approach of #1602, we ignore the region in the
+    // user's locale.
+    const userLocale = Services.appShell.hiddenDOMWindow.navigator.language.split('-')[0];
+    return x.locale_grantlist.includes(userLocale);
+  });
```
Comment (pdehaan) [inline addon/lib/toolbar-button.js:100]: Do we need to check the `locale_blocklist` as well? [/docs/content/template.yaml:94-99](https://github.com/mozilla/testpilot/blob/97fdd0735024f85014aa8645ef0dc763a8aadcf7/docs/content/template.yaml#L94-L99)

([Ref](https://github.com/mozilla/testpilot/blob/2218655b4ecc33a53dfc42d2cc4de91d3a45047e/docs/content/reference.md#locales))

Comment (jaredhirsch) [reply addon/lib/toolbar-button.js:100]: @pdehaan yup! that's #1967, pushing a fix rn

**Comment by jaredhirsch** [comment]
OK, updated. Ready for review

**Comment by jaredhirsch** [comment]
With `navigator.language` set to `de-ch`, Activity Stream and NoMore404s are correctly hidden, and Cliqz is correctly shown:

screen shot 2017-01-04 at 5 07 42 pm

With `navigator.language` set to `en-US`, Cliqz is correctly hidden, and Activity Stream shown:

screen shot 2017-01-04 at 5 08 01 pm

**Comment by fzzzy** [comment]
Looks sane to me. r+

File: addon/lib/toolbar-button.js [hunk 2]
```diff
@@ -68,9 +72,18 @@ function getExperimentList(availableExperiments, installedAddons) {
...
     return experiment;
   })
-  .filter(x => x.active || !x.completed); // remove inactive, completed experiments
-
-  // Sort new experiments to the top, otherwise sort by reverse-chronological
+  // Remove inactive, completed experiments.
+  .filter(x => x.active || !x.completed)
+  // Remove experiments with a locale_grantlist, unless the user's primary
+  // locale is in the grantlist. #1961
+  .filter(x => !('locale_grantlist' in x) || x.locale_grantlist.includes(userLocale))
```
Comment (fzzzy) [inline addon/lib/toolbar-button.js]: In python I would prefer to write this x.get('locale_grantlist', []).includes(userLocale) but I'm not sure how to write that in js.

**Review by ckprice — DISMISSED** [review]
@6a68 can you confirm if this comment was addressed in #1961 

> The "New" badge is also displayed for Cliqz experiment on other locales than DE.

I pulled down your branch and was still seeing the "New" badge (presumably for Cliqz) while configured for "en-US". Could have just been some crud with my set up though.

File: addon/lib/toolbar-button.js (hunk 2 above)
Comment (jaredhirsch) [reply addon/lib/toolbar-button.js]: Oh, that's a great point! You were actually really close:

```js
(x.locale_grantlist || []).includes(userLocale)
```

But...I'm just going to rewrite this to be less terse. Looking at it this morning, I think a simpler style will invite contributions from less-experienced devs, and decrease the chances of making a silly mistake due to misreading something when tired.

**Comment by jaredhirsch** [comment]
@ckprice I didn't fix the 'new' badge issue in this PR, but I could add it

**Comment by jaredhirsch** [comment]
@fzzzy Pushed some less cryptic code :+1:

@ckprice I'll add the 'new' filtering now. Looks like it's been filed as #1973.

**Comment by jaredhirsch** [comment]
Turns out there's another bug here, in the notification code: notifications are triggered if a new experiment is available, without filtering on important stuff like locale lists or launch date.

So, I'm going to pull out the filtering code, abstract `store.availableExperiments` behind a storage getter that dynamically recalculates the available experiments, and that should keep all the gnarly filtering code from cluttering the UI code.

**Comment by jaredhirsch** [comment]
Actually, I'm going to hold any additional changes (which includes some refactoring) for a future PR. Let's get these blocker bugs, excluding the 'new' label being localized, landed and QA verified. This should help keep tensions low going into the weekend.

**Pull request merged**